File: python/scripts/gibbs_sampling.py
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path
from scipy import stats
import pandas as pd
import time


def gibbs_lm(
    y, X, n_iter=10000, warmup=2000, n_chains=4,
    b0=None, B0=None, a0=0.01, d0=0.01, seed=123
):
    """
    Gibbs sampler for Bayesian linear regression with conjugate priors.
    Uses Normal-Inverse-Gamma setup so we can sample beta and sigma2 in closed form.
    """

    y = np.asarray(y).reshape(-1)
    X = np.asarray(X)
    n = y.shape[0]
    p = X.shape[1]

    if b0 is None:
        b0 = np.zeros(p)
    if B0 is None:
        B0 = np.eye(p) * 1e-4

    b0 = np.asarray(b0).reshape(-1)
    B0 = np.asarray(B0)

    chain_results = []

    # stuff that doesn't change between iterations
    XtX = X.T @ X
    Xty = X.T @ y
    Bn = XtX + B0
    bn = np.linalg.solve(Bn, Xty + B0 @ b0)
    L = np.linalg.cholesky(Bn)  # for sampling from multivariate normal

    for chain in range(n_chains):
        rng = np.random.default_rng(seed + chain)

        beta_store = np.zeros((n_iter, p))
        sigma2_store = np.zeros(n_iter)

        beta_draw = np.zeros(p)
        sigma2_draw = 1.0

        t0 = time.time()

        for iter_idx in range(n_iter):
            # sample beta from its full conditional
            z = rng.standard_normal(p)
            v = np.linalg.solve(L, z)
            w = np.linalg.solve(L.T, v)
            beta_draw = bn + np.sqrt(sigma2_draw) * w

            # sample sigma2 from its full conditional
            resid = y - X @ beta_draw
            an = a0 + n / 2.0
            quad_prior = (beta_draw - b0).T @ B0 @ (beta_draw - b0)  # prior term matters since beta is sigma2-scaled
            dn = d0 + 0.5 * (resid @ resid + quad_prior)
            sigma2_draw = 1.0 / rng.gamma(shape=an, scale=1.0 / dn)

            beta_store[iter_idx, :] = beta_draw
            sigma2_store[iter_idx] = sigma2_draw

        elapsed = time.time() - t0

        chain_results.append({
            "beta": beta_store[warmup:, :],
            "sigma2": sigma2_store[warmup:],
            "time_elapsed": elapsed
        })

        print(f"Chain {chain + 1} completed. time_elapsed={elapsed:.3f}s")

    return chain_results


def beta_trace_plot(beta_list, model_name, plot_dir='../outputs'):
    n_chains = len(beta_list)
    p = beta_list[0].shape[1]
    output_dir = Path(plot_dir) / model_name
    output_dir.mkdir(parents=True, exist_ok=True)
    colors = plt.cm.rainbow(np.linspace(0, 1, n_chains))
    
    for j in range(p):
        fig, ax = plt.subplots(figsize=(10, 6))
        for chain_idx in range(n_chains):
            iterations = np.arange(1, len(beta_list[chain_idx][:, j]) + 1)
            ax.plot(iterations, beta_list[chain_idx][:, j], 
                   color=colors[chain_idx], alpha=0.7, 
                   label=f'Chain {chain_idx + 1}', linewidth=0.5)
        
        ax.set_xlabel('Iteration', fontsize=12)
        ax.set_ylabel(rf'$\beta_{{{j}}}$', fontsize=12)
        ax.set_title(rf'Trace plot: $\beta_{{{j}}}$', fontsize=14)
        ax.legend(loc='upper right')
        ax.grid(True, alpha=0.3)
        
        filename = output_dir / f'beta_trace_{j+1}.png'
        plt.savefig(filename, dpi=150, bbox_inches='tight')
        plt.close()
        print(f"Saved: {filename}")


def sigma2_trace_plot(sigma2_list, model_name, plot_dir='../outputs'):
    n_chains = len(sigma2_list)
    output_dir = Path(plot_dir) / model_name
    output_dir.mkdir(parents=True, exist_ok=True)
    colors = plt.cm.rainbow(np.linspace(0, 1, n_chains))
    
    fig, ax = plt.subplots(figsize=(10, 6))
    for chain_idx in range(n_chains):
        iterations = np.arange(1, len(sigma2_list[chain_idx]) + 1)
        ax.plot(iterations, sigma2_list[chain_idx], 
               color=colors[chain_idx], alpha=0.7,
               label=f'Chain {chain_idx + 1}', linewidth=1.5)
    
    ax.set_xlabel('Iteration', fontsize=12)
    ax.set_ylabel(r'$\sigma^2$', fontsize=12)
    ax.set_title(r'Trace plot: $\sigma^2$', fontsize=14)
    ax.legend(loc='upper right')
    ax.grid(True, alpha=0.3)
    
    filename = output_dir / 'Sigma_trace.png'
    plt.savefig(filename, dpi=150, bbox_inches='tight')
    plt.close()
    print(f"Saved: {filename}")


def beta_summary_stats(beta_list):
    beta_all = np.vstack(beta_list)
    summary = {
        'Mean': np.mean(beta_all, axis=0),
        'SD': np.std(beta_all, axis=0, ddof=1),
        '2.5%': np.percentile(beta_all, 2.5, axis=0),
        '97.5%': np.percentile(beta_all, 97.5, axis=0)
    }
    df_summary = pd.DataFrame(summary).round(4)
    print("\nBeta Summary Statistics:")
    print(df_summary)
    return df_summary


def sigma2_summary_stats(sigma2_list):
    sigma2_all = np.concatenate(sigma2_list)
    
    summary = {
        'Mean': np.mean(sigma2_all),
        'SD': np.std(sigma2_all, ddof=1),
        '2.5%': np.percentile(sigma2_all, 2.5),
        '97.5%': np.percentile(sigma2_all, 97.5)
    }
    
    summary = {k: round(v, 4) for k, v in summary.items()}
    
    print("\nSigma^2 Summary Statistics:")
    for key, value in summary.items():
        print(f"  {key}: {value}")
    
    return summary


if __name__ == "__main__":
    print("=" * 60)
    print("Testing Gibbs sampler")
    print("=" * 60)
    
    np.random.seed(42)
    n = 100
    p = 3
    
    X = np.column_stack([np.ones(n), np.random.randn(n, p-1)])
    true_beta = np.array([2.0, 1.5, -1.0])
    true_sigma2 = 1.0
    
    y = X @ true_beta + np.random.randn(n) * np.sqrt(true_sigma2)
    
    print(f"\nTest data: n = {n}, p = {p}")
    print(f"True beta: {true_beta}")
    print(f"True sigma^2: {true_sigma2}")
    
    print("\nRunning sampler...")
    results = gibbs_lm(y, X, n_iter=5000, warmup=1000, n_chains=3)
    
    beta_list = [chain['beta'] for chain in results]
    sigma2_list = [chain['sigma2'] for chain in results]
    
    beta_summary = beta_summary_stats(beta_list)
    sigma2_summary = sigma2_summary_stats(sigma2_list)
    
    print("\nDone.")

File: r_final/scripts/Gibbs_Sampling.R

library(MASS)

#' @param y Numeric vector. Response variable.
#' @param X Numeric matrix or data frame. Predictor variables.
#' @param n_iter Integer. Total number of Gibbs sampling iterations. Default is 10000.
#' @param warmup Integer. Number of initial samples to discard (burn-in). Default is 2000.
#' @param n_chains Integer. Number of independent MCMC chains to run. Default is 3.
#' @param b0 Numeric vector. Prior mean for regression coefficients (\eqn{\beta}). Default is a zero vector.
#' @param B0 Numeric matrix. Prior precision matrix for regression coefficients (\eqn{\beta}). 
#'   Default is a diagonal matrix with small values (1e-4) indicating weak prior information.
#' @param a0 Numeric. Shape parameter for inverse gamma prior on \eqn{\sigma^2}. Default is 0.01.
#' @param d0 Numeric. Scale parameter for inverse gamma prior on \eqn{\sigma^2}. Default is 0.01.
#' @param seed Integer. Random seed base value for reproducibility across chains. Default is 123.
#
#' @return A list containing posterior samples from each chain:
#' \describe{
#'   \item{chain_results}{A list of length equal to `n_chains`, where each element contains:}
#'     \describe{
#'       \item{beta}{Matrix of sampled regression coefficients after burn-in; rows correspond to iterations, columns to predictors.}
#'       \item{sigma2}{Vector of sampled error variances after burn-in for that chain.}
#'     }
#' }
#'

gibbs_lm <- function(y, X,
                     n_iter = 10000,
                     warmup = 2000,
                     n_chains = 3,
                     b0 = NULL,
                     B0 = NULL,
                     a0 = 0.01,
                     d0 = 0.01,
                     seed = 123) {
  
  

  
  n <- length(y) # number of observations
  p <- ncol(X)   # number of predictors 
  
  

  # -----------------------------
  # Default priors
  # -----------------------------
  
  if (is.null(b0)) b0 <- rep(0, p)
  if (is.null(B0)) B0 <- diag(1e-4, p)
  
  # Initialize chain results
  chain_results <- vector("list", n_chains)  
  
  
  for (chain in seq_len(n_chains)) {
    
  set.seed(seed + chain)
    
  # Storage for samples
  beta_store   <- matrix(NA, n_iter, p)
  sigma2_store <- numeric(n_iter)
  
  

  # Initialize parameters
  beta_draw  <- rep(0, p)
  sigma2_draw <- 1
  
  
  # Pre-compute for efficiency
  XtX <- t(X) %*% X
  Xty <- t(X) %*% y
  

  # Gibbs sampling loop
  
  Bn <- XtX + B0         
  bn <- solve(Bn, Xty + B0 %*% b0)
  
  for (iter in 1:n_iter) {
    
    # ----- Sample beta given sigma^2 -----
    
    beta_draw <- mvrnorm(
      n = 1,
      mu = bn,
      Sigma = sigma2_draw * solve(Bn)
    )
    
    # ----- Sample sigma^2 given beta -----
    
    resid <- y - X %*% beta_draw
    a_n <- a0 + n / 2
    
    quad_prior <- t(beta_draw - b0) %*% B0 %*% (beta_draw - b0)
    
    d_n <- d0 + 0.5 * (sum(resid^2) + as.numeric(quad_prior))
    
    sigma2_draw <- 1 / rgamma(1, shape = a_n, rate = d_n)
    
    # --- ------ Store samples -----------
    
    beta_store[iter, ]  <- beta_draw
    sigma2_store[iter]  <- sigma2_draw
  }
  
  # --------------------------------------
  # Return posterior samples after burn-in
  # --------------------------------------
  
  chain_results[[chain]] <- list(
    beta   = beta_store[-(1:warmup), ],
    sigma2 = sigma2_store[-(1:warmup)]
  )
  
  cat("Chain", chain, "completed.\n")
  }
  
  return(chain_results)
}

#-----------------------------------------
# Trace Plots for Beta & Sigma^2
#-----------------------------------------

beta_trace_plot <- function(beta_list, model_name) {
  
  n_chains <- length(beta_list)
  p <- ncol(beta_list[[1]])

  plot_dir <- file.path("..", "outputs", model_name, "plots", "trace")
  
  if (!dir.exists(plot_dir)) {
    dir.create(plot_dir, recursive = TRUE)
    message("Created folder: ", plot_dir)
  }
  
  colors <- rainbow(n_chains)
  
  for (j in 1:p) {
    
    filename <- file.path(plot_dir, paste0("beta_trace_", j, ".png"))
    png(filename = filename, width = 1600, height = 1200, res = 150)
    plot(NULL,
         xlim = c(1, nrow(beta_list[[1]])),
         ylim = range(sapply(beta_list, function(x) x[, j])),
         type = "n",
         main = bquote("Trace plot: " ~ beta[.(j)]),
         xlab = "Iteration",
         ylab = bquote(beta[.(j)]))
    
    for (chain in 1:n_chains) {
      lines(beta_list[[chain]][, j], col = colors[chain])
    }
    legend("topright", legend = paste("Chain", 1:n_chains),
           col = colors, lty = 1, cex = 0.8)
    
    dev.off()
    message("Saved: ", filename)
  }
  
}


sigma2_trace_plot <- function(sigma2_list,model_name){
  
  n_chains <- length(sigma2_list)
  colors <- rainbow(n_chains)
  
  plot_dir <- file.path("..", "outputs", model_name, "plots", "trace")
  
  if (!dir.exists(plot_dir)) {
    dir.create(plot_dir, recursive = TRUE)
    message("Created folder: ", plot_dir)
  }
  
  filename <- file.path(plot_dir, paste0("Sigma_trace.png"))
  png(filename = filename, width = 1600, height = 1200, res = 150)
  plot(NULL,
       xlim = c(1, length(sigma2_list[[1]])),
       ylim = range(sapply(sigma2_list, range)),
       type = "n",
       main = expression("Trace plot: " ~ sigma^2),
       xlab = "Iteration",
       ylab = expression(sigma^2))
  
  # Plot each chain
  for (chain in 1:n_chains) {
    lines(sigma2_list[[chain]], col = colors[chain], lwd = 1.5)
  }
  
  legend("topright", legend = paste("Chain", 1:n_chains),
         col = colors, lty = 1, cex = 0.8)
  dev.off()
  message("Saved: ", filename)
}

#---------------------------------------------
# Posterior Summary Statistics
#---------------------------------------------

beta_summary_stats <- function(beta_list) {
  
  #Combine all chains
  beta_all <- do.call(rbind, beta_list)
  beta_summary <- apply(beta_all, 2, function(x) {
    c(
      mean = mean(x),
      sd   = sd(x),
      q2.5 = quantile(x, 0.025),
      q97.5 = quantile(x, 0.975)
    )
  })
  
  # Transpose + add column names
  beta_summary <- t(beta_summary)
  colnames(beta_summary) <- c("Mean", "SD", "2.5%", "97.5%")
  
  return(round(beta_summary, 4))
}


sigma2_summary_stats <- function(sigma2_list) {
  
  sigma2_all <- unlist(sigma2_list)
  sigma2_summary <- c(
    mean = mean(sigma2_all),
    sd   = sd(sigma2_all),
    q2.5 = quantile(sigma2_all, 0.025),
    q97.5 = quantile(sigma2_all, 0.975)
  )
  
  return(round(sigma2_summary, 4))
}

